File: python/models/event.py
import dataclasses
import datetime
import typing

from google.cloud import firestore

from .comment import CommentModel


@dataclasses.dataclass(kw_only=True)
class EventModel:
    name: str = ""
    photo: str = ""
    start_date: datetime.datetime = datetime.datetime.now()
    end_date: datetime.datetime = datetime.datetime.now()
    createdBy: firestore.DocumentReference = None
    comments: typing.List[CommentModel] = dataclasses.field(default_factory=list)

    reference: typing.Optional[firestore.DocumentReference]

    def to_map(self):
        data = dataclasses.asdict(self)
        del data["reference"]

        return data

File: models/event.ts
import type { DocumentData, DocumentReference } from 'firebase/firestore'
import { Timestamp } from 'firebase/firestore'
import type { IComment } from './comment'

export interface IEvent {
  name: string
  photo: string
  startDate: Date | Timestamp
  endDate: Date | Timestamp
  createdBy: DocumentReference
  comments: IComment[]
}

export class EventModel implements IEvent {
  name: string
  photo: string
  startDate: Date | Timestamp
  endDate: Date | Timestamp
  createdBy: DocumentReference
  comments: IComment[]

  reference: DocumentReference | null

  constructor(data: IEvent, reference: DocumentReference | null) {
    this.name = data.name || ''
    this.photo = data.photo || ''
    this.startDate = data.startDate instanceof Timestamp ? data.startDate.toDate() : data.startDate
    this.endDate = data.endDate instanceof Timestamp ? data.endDate.toDate() : data.endDate
    this.createdBy = data.createdBy
    this.comments = data.comments || []

    this.reference = reference
  }

  toMap() {
    return {
      name: this.name,
      photo: this.photo,
      startDate: this.startDate,
      endDate: this.endDate,
      createdBy: this.createdBy,
      comments: this.comments,
    }
  }
}

export function mapEvent(event: DocumentData) {
  return new EventModel(
    event.data(),
    event.ref,
  )
}
